Connect windows webauthn with tsh

tobiaszheller opened the pull request:
implementes https://github.com/gravitational/teleport/pull/16360
Uses wrapper implemented in https://github.com/gravitational/teleport/pull/16634 to connec via tsh.

What has changed?

- support for fido devices and windows hello on windows platform
- storing mfa-mode in tsh profile file

Support for extension: app_id will be provided in other PR.
Integration/manual tests will be provided in other PR.
Deleting credentials will be added in separate PR (only supported in v4)

Solves: https://github.com/gravitational/teleport/issues/15148

codingllama commented on lib/auth/webauthncli/api.go:
```diff
@@ -138,14 +151,32 @@ func crossPlatformLogin(
 	}
 
+	if winwebauthn.IsAvailable() {
+		log.Debug("WIN_WEBAUTHN: Using windows webauthn for credential creation")
```
```suggestion
		log.Debug("WIN_WEBAUTHN: Using windows webauthn for credential assertion")
```

codingllama commented on lib/auth/webauthncli/api.go:
```diff
@@ -116,10 +129,10 @@ func Login(
 		return crossPlatformLogin(ctx, origin, assertion, prompt, opts)
```
Since the Windows code handles both platform and cross-platform, could we pull the "if winwebauthn.IsAvailable()" before this block and handle all Windows code there?

tobiaszheller replied:
done

codingllama commented on lib/auth/webauthncli/api.go:
```diff
@@ -138,14 +151,32 @@ func crossPlatformLogin(
 
+	if winwebauthn.IsAvailable() {
+		log.Debug("WIN_WEBAUTHN: Using windows webauthn for credential creation")
+		inOpts := &winwebauthn.LoginOpts{}
```
Attachment here would always be cross-platform. Similarly in the method below.

tobiaszheller replied:
I will move that `webauthnwin.Login` above platform/crossplatform code so this will simplify

codingllama commented on lib/auth/webauthncli/api.go:
```diff
@@ -184,6 +215,11 @@ func Register(
 		return FIDO2Register(ctx, origin, cc, prompt)
 	}
 
+	if winwebauthn.IsAvailable() {
```
Move this before FIDO2? It seems like we'd like it to have preference regardless.

codingllama commented on lib/auth/webauthncli/platform_windows.go:
```diff
@@ -17,5 +17,5 @@ package webauthncli
 // WebAuthn-compatible logins.
 func HasPlatformSupport() bool {
-	return false
+	return true
```
Maybe?

```suggestion
	return winwebauthn.IsAvailable()
```

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+// limitations under the License.
+
+// Package winwebauthn is wrapper around Windows webauthn API.
+// It loads system webauthn.dll and uses it's method.
```
```suggestion
// It loads system webauthn.dll and uses its methods.
```

I suppose?

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+
+// Package winwebauthn is wrapper around Windows webauthn API.
+// It loads system webauthn.dll and uses it's method.
+// It supports API versions 1-4.
```
This possibly becomes 1+ if some of my comments in the parent PR are correct.

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+// Login implements Login for Windows Webauthn API.
+// It returns an MFAAuthenticateResponse.
```
Ditch? It's not adding much.

```suggestion
```

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+	}
+	resp, err := native.GetAssertion(origin, assertion.Response, opts)
+	if err != nil {
+		// TODO(tobiaszheller): proper error
```
What do you mean?

tobiaszheller replied:
So far I am not using any typed errors (I am just returning whatever error message is returned from API) but maybe it's good for now because I am not doing any error checks and fallback logic?

codingllama replied:
Gotcha. Let's clarify the TODO so a random person could understand what you meant (and maybe even fix it!)

tobiaszheller replied:
done

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+	resp, err := native.GetAssertion(origin, assertion.Response, opts)
+	if err != nil {
+		// TODO(tobiaszheller): proper error
+		return nil, "", err
```
```suggestion
		return nil, "", trace.Wrap(err)
```

Same where applicable.

codingllama replied:
Bump so it won't be forgotten.

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+
+	resp, err := native.MakeCredential(origin, cc.Response)
+	if err != nil {
+		// TODO(tobiaszheller): proper error
```
Ditto.

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+	HasCompileSupport bool
+	IsAvailable       bool
+	HasPlatformUV     bool
+	APIVersion        int
```
```suggestion
	WebAuthnNDLLVersion        int
```

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+
+// Login implements Login for Windows Webauthn API.
+// It returns an MFAAuthenticateResponse.
+// Most callers should call Login directly, as it is correctly guarded by
```
> Most callers should call Login directly

This doesn't seem applicable here, what else would they call?

codingllama commented on lib/auth/winwebauthn/api.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+// Register implements Register for Windows Webauthn API.
+// Most callers should call Register directly, as it is correctly guarded by
```
Ditto.

codingllama commented on tool/tsh/mfa.go:
```diff
@@ -207,7 +208,7 @@ func newMFAAddCommand(parent *kingpin.CmdClause) *mfaAddCommand {
 	c.Flag("type", fmt.Sprintf("Type of the new MFA device (%s)", strings.Join(defaultDeviceTypes, ", "))).
 		EnumVar(&c.devType, defaultDeviceTypes...)
-	if wancli.IsFIDO2Available() {
+	if wancli.IsFIDO2Available() || winwebauthn.IsAvailable() {
```
Do we want to add a higher-level check to wancli? Should IsFIDO2Available return true if the Windows impl is present? Strictly speaking if we have winwebauthn we do have FIDO2.

tobiaszheller replied:
that's good question, I was also thinking about some higher-level check but did not come with good name. Do you have any good suggestions? `wancli.IsPasswordlessCapableClient`?

codingllama replied:
Looking at the usages, I think we could do 2 separate methods:

```go
package webauthncli

func isLibfido2Enabled() bool {...} // same as the "old" IsFIDO2Available, now private

func IsFIDO2Available() bool { return isLibfido2Enabled() || winwebauthn.IsAvailable() }
```

Touch ID remains as a separate check, as it's a separate device type.

tobiaszheller replied:
I am not sure if we can make IsFIDO2Available change how it works. It is used in other places which clearly works only with libfido2 library. 
Maybe we can done something like that:
```
IsFIDO2Available  ->> IsLibfido2Available
IsPasswordlessSelectable() bool { return  IsLibfido2Available || winwebauthn.IsAvailable()}
```

tobiaszheller replied:
good idea, i misread first comment but now it's clear. Changes already pushed.

codingllama reviewed (commented):
Just taking a quick glance, will review tests / CLI more closely once we get the base going.

This PRs isn't too big, but it's doing a few separate jobs:

1. Adding a public API (and tests) to winwebauthn
2. Wiring winwebauthn to webauthncli (and thus tsh)
3. Adding --mfa-mode to tsh profile
4. Adding debugging commands to tsh

In the future consider breaking into smaller PRs, or at least breaking the commits a bit so different changes are not mixed together.

codingllama commented on lib/auth/webauthnwin/api_test.go:
```diff
@@ -0,0 +1,255 @@
+	}
+}
+
+func (f fakeNative) GetAssertion(origin string, in *wanlib.CredentialAssertion, loginOpts *webauthnwin.LoginOpts) (*wanlib.CredentialAssertionResponse, error) {
```
Let's write a happy-path test for login and registration. It's not much, but I'd rather have one.

codingllama commented on lib/auth/webauthnwin/api_test.go:
```diff
@@ -0,0 +1,255 @@
+			wantErr: "user display name",
+		},
+		{
+			name:   "rrk nil user ID",
```
Most of the tests here are uninteresting, I think. I'd rather refactor/reuse the validations, have that tested elsewhere, and just make sure we test that the validations are "plugged in" here, plus whatever unique behaviors we have.

Ditto for registration.

codingllama reviewed (commented):
As said in #16634, I think we are leaving too much behind `windows` build tags, which most developers (and CI) will never check.

I suggest pulling most of the struct conversions out of the `windows` tag and making the native interface more Windows-specific. This way we can fake better and cover more code with automated testing.

codingllama commented on lib/auth/webauthnwin/api.go:
```diff
@@ -33,10 +45,174 @@ const (
+		return nil, "", trace.BadParameter("assertion required")
+	case len(assertion.Response.Challenge) == 0:
+		return nil, "", trace.BadParameter("assertion challenge required")
+	case assertion.Response.RelyingPartyID == "":
```
Btw, let's do this in a separate PR so we can backport the changes.

tobiaszheller replied:
i will check it, i wanted to move windows specific required check here also but I can check if most of it is common and migrate it to common fn

tobiaszheller replied:
https://github.com/gravitational/teleport/pull/16879/
here is pr

codingllama replied:
Great, thanks!

tobiaszheller commented on tool/tsh/winwebauthn.go:
```diff
@@ -0,0 +1,87 @@
+func newWebauthnwinCommand(app *kingpin.Application) *webauthnwinCommand {
+	wid := app.Command("webauthnwin", "Manage Windows WebAuthn").Hidden()
+	cmd := &webauthnwinCommand{
+		support:     newWebauthnwinSupportCommand(wid),
+		diagnostics: newWebauthnwinDiagnosticsCommand(wid),
```
one works without user interactions, but maybe after all that can be 1 command

codingllama replied:
I think that's alright, `fido2 diag` requires interaction too. I'd rather have a single diag command than 2, it's easier to ask people to do a single command.

tobiaszheller replied:
reworked, ptaL

tobiaszheller commented on lib/auth/webauthncli/platform_windows.go:
```diff
@@ -14,8 +14,10 @@
 // WebAuthn-compatible logins.
 func HasPlatformSupport() bool {
-	return false
+	return webauthnwin.IsAvailable()
```
done

tobiaszheller commented on lib/auth/webauthnwin/api.go:
```diff
@@ -33,10 +45,174 @@ const (
+	}
+	resp, err := native.GetAssertion(origin, assertion, opts)
+	if err != nil {
+		// TODO(tobiaszheller): right now error directly from webauthn.dll is
+		// returned. At some point probably we want to introducde typed errors.
```
do you think it's worth investing time into those typed errors now?

codingllama replied:
I think we'll never follow up on this, so I'd be OK deleting the TODO and moving on. I'm not sure why we would want typed errors.

tobiaszheller commented:
@codingllama here is commit which moves part of windows outside so we can add more UT: https://github.com/gravitational/teleport/pull/16635/commits/f2a90c65bf4cdec052522d4df8f325e92b864e6e
let me know what you thing about that kind of tests? is it what you were thinking about? 
Basically we can assert many incoming parameters to "pure windows" part. 
I will extend test cases more later but wanted to check if I going correct path

tobiaszheller commented on lib/auth/webauthncli/api.go:
```diff
@@ -189,3 +214,15 @@ func Register(
+
+// HasPlatformSupport returns true if the platform supports client-side
+// WebAuthn-compatible logins.
+func HasPlatformSupport() bool {
+	return IsFIDO2Available() || touchid.IsAvailable()
+}
```
@codingllama looking at it from outside, you cannot see that it can also be suypported from windows hello, maybe it should be after all something like that?
```suggestion
func HasPlatformSupport() bool {
	return isLibfido2Enabled() || webauthnwin.IsAvailable() || touchid.IsAvailable()
}
```

codingllama replied:
I would go with the 1st, so eventual changes to IsFIDO2Available reflect automatically here.

codingllama commented:
> @codingllama here is commit which moves part of windows outside so we can add more UT: [f2a90c6](https://github.com/gravitational/teleport/commit/f2a90c65bf4cdec052522d4df8f325e92b864e6e) let me know what you thing about that kind of tests? is it what you were thinking about? Basically we can assert many incoming parameters to "pure windows" part. I will extend test cases more later but wanted to check if I going correct path

Yep, that's more like it. We get a bit more coverage and the native parts are as thin as can be. +1.

tobiaszheller commented:
@codingllama PTAL, I have applied most of comments. Also I am not sure how deep we want to go with UT, i have convered so far 3 fields there, UV, RRK, and Attachment.

tobiaszheller commented on lib/auth/webauthnwin/ctypes.go:
```diff
@@ -61,9 +59,11 @@ type webauthnAuthenticatorMakeCredentialOptions struct {
 	dwTimeoutMilliseconds uint32
 	// Credentials used for exclusion.
-	CredentialList webauthnCredentials
+	//nolint:unused
```
i needed to add that //nolint:unused annotations. This types reflect API types however few fields are not used. Also linter does not run inside windows build tags, that's why some incorrectly reported as unused.

tobiaszheller commented on lib/auth/webauthnwin/api_test.go:
```diff
@@ -0,0 +1,250 @@
+			origin:   origin,
+			createCC: func() *wanlib.CredentialCreation { return okCC },
+			assertFn: func(t *testing.T, ccr *webauthn.CredentialCreationResponse, req *makeCredentialRequest) {
+				assert.Equal(t, uint32(0), req.opts.dwAuthenticatorAttachment)
```
done

codingllama replied:
Much better, thanks!

tobiaszheller commented on lib/auth/webauthnwin/api_test.go:
```diff
@@ -0,0 +1,250 @@
+				assert.Equal(t, uint32(6), req.opts.dwVersion)
+				assert.Equal(t, uint32(3), req.opts.dwUserVerificationRequirement)
+				assert.Equal(t, uint32(0), req.opts.dwAuthenticatorAttachment)
+
```
PTAL, I am not sure If i done it correctly.

codingllama replied:
Sorry, I meant removing the extra blank lines. Nevermind this, let's move on.

tobiaszheller commented:
@codingllama adding `windows` build tag seems to no be trivial tasks. When I tried running it manually it throws 1k+ errors. 
I can take a look later at linter support for windows and add it in separate PR.
For now I have added more docs to `nolint` and removed few of them because I have also removed some of the fields (just keep them to have correct struct size).

tobiaszheller commented on lib/auth/webauthnwin/api.go:
```diff
@@ -33,10 +45,174 @@ const (
+		return nil, trace.BadParameter("credential creation relying party ID required")
+	}
+
+	rrk := cc.Response.AuthenticatorSelection.RequireResidentKey != nil && *cc.Response.AuthenticatorSelection.RequireResidentKey
```
it's also in https://github.com/gravitational/teleport/pull/16879/

codingllama commented on lib/auth/webauthnwin/ctypes.go:
```diff
@@ -108,28 +124,34 @@ type webauthnAuthenticatorMakeCredentialOptions struct {
 }
+
+//nolint:unused // This struct is kept just to keep size of struct valid.
 type webauthnExtenstions struct {
```
```suggestion
type webauthnExtensions struct {
```

codingllama reviewed (commented):
(Waiting for diffbase and #16879 to land.)

codingllama commented on lib/auth/webauthnwin/api.go:
```diff
@@ -33,10 +44,207 @@ const (
+				DisplayName: "test",
+			},
+			Parameters: []protocol.CredentialParameter{
+				{
+					Type:      protocol.PublicKeyCredentialType,
+					Algorithm: webauthncose.AlgRS256,
+				},
```
```suggestion
				{
					Type:      protocol.PublicKeyCredentialType,
					Algorithm: webauthncose.AlgES256,
				},
				{
					Type:      protocol.PublicKeyCredentialType,
					Algorithm: webauthncose.AlgRS256,
				},
```

In case someone wants to use a FIDO2/U2F device.

Commit:
Windows Webauthn add Login/Register/Diag methods

Commit:
Windows Webauthn connect with webauthncli package

Commit:
Windows Webauthn add diag command to tsh

Commit:
Store mfa-mode in profile

tobiaszheller commented:
@xacrimon @zmb3 could you take a look? It has grown quite a big unfortunately. It is divided into 4 commits (hope it makes it a little bit easier to review).

codingllama commented on lib/auth/webauthncli/export.go:
```diff
@@ -0,0 +1,18 @@
+package webauthncli
+
+// HasPlatformSupport provides indirections for tests.
+var HasPlatformSupport = hasPlatformSupport
```
I wouldn't do this here, push the var to lib/client - that's where we need it mocked. lib/auth/webauthncli should just expose a regular method, as before. This cuts the scope of the "mocking", so it's a bit more safe.

tobiaszheller replied:
https://github.com/gravitational/teleport/pull/16635/commits/a9ebb25fdec5f199878bca52698ff3bc1ae61960
i have done it from scratch, PTAL

codingllama replied:
Thanks!

codingllama commented on lib/auth/webauthncli/api.go:
```diff
@@ -215,10 +215,10 @@ func Register(
 // WebAuthn-compatible logins.
-func HasPlatformSupport() bool {
-	return IsFIDO2Available() || touchid.IsAvailable()
+func hasPlatformSupport() bool {
```
I would keep it public, see the other comment.

codingllama commented on lib/client/api_login_test.go:
```diff
@@ -77,9 +77,17 @@ func TestTeleportClient_Login_local(t *testing.T) {
 	// Reset functions after tests.
 	oldStdin, oldWebauthn := prompt.Stdin(), *client.PromptWebauthn
+	oldHasPlatformSupport := wancli.HasPlatformSupport
+	t.Cleanup(func() {
```
I'm a bit confused, is this supposed to be wrapped in a t.Cleanup?

tobiaszheller replied:
no, I guess it was some copy paste mistake :(

tobiaszheller commented on lib/auth/webauthnwin/api.go:
```diff
@@ -33,10 +44,211 @@ const (
 )
 
+// nativeWebauthn represents the native windows webauthn interface.
+// Implementors must provide a global variable called `native`.
```
Interface is defined in common file. 
[Further in codebase](https://github.com/gravitational/teleport/blob/d60125c816da6e6eec74eaa78e79e57cbb208999/lib/auth/webauthnwin/api.go#L93) we use `native` variable which implements given interface, example: 
```
func Login() error {
...
resp, err := native.GetAssertion()
}
``` 
and native is added in both [windows](https://github.com/gravitational/teleport/blob/d60125c816da6e6eec74eaa78e79e57cbb208999/lib/auth/webauthnwin/webauthn_windows.go#L48) and [non_windows](https://github.com/gravitational/teleport/blob/d60125c816da6e6eec74eaa78e79e57cbb208999/lib/auth/webauthnwin/webauthn_other.go#L26) build tags.
Usage of interface allows us to mock it during UT. 


Do you have suggestions how to document it in more clear way?
```
Package variable `native` must be used to provide implementation of that interface.
For windows platform it will be real implementation and for other noop.
```

zmb3 replied:
I see now. I think it's fine as-is.

tobiaszheller commented:
@zmb3 PTAL
